pahfit/model.py

```diff
@@ -166,11 +170,111 @@ def guess(self, spec: Spectrum1D, redshift=None):
+
+        # simple linear interpolation function for spectrum
+        sp = interpolate.interp1d(xz, yz)
+
+        # we will repeat this loop logic several times
+        def loop_over_non_fixed(kind, parameter, estimate_function, force=False):
+            for row_index in np.where(self.features["kind"] == kind)[0]:
+                row = self.features[row_index]
+                if not bounded_is_fixed(row[parameter]) or force:
+                    guess_value = estimate_function(row)
+                    # print(f"{row['name']}: setting {parameter} to {guess_value}")
+                    self.features[row_index][parameter][0] = guess_value
+
+        # guess starting point of bb
+        def starlight_guess(row):
+            bb = BlackBody1D(1, row["temperature"][0])
+            w = wmin + 0.1  # the wavelength used to compare
+            if w < 5:
+                # wavelength is short enough to not have numerical
+                # issues. Evaluate both at w.
+                amp_guess = sp(w) / bb(w)
+            else:
+                # wavelength too long for stellar BB. Evaluate BB at
+                # 5 micron, and spectrum data at minimum wavelength.
+                wsafe = 5
+                amp_guess = sp(w) / bb(wsafe)
+
+            return amp_guess
+
+        loop_over_non_fixed("starlight", "tau", starlight_guess)
+
+        # count number of blackbodies in the model
+        nbb = len(self.features[self.features["kind"] == "dust_continuum"])
+
+        def dust_continuum_guess(row):
+            temp = row["temperature"][0]
+            fmax_lam = 2898.0 / temp
+            bb = BlackBody1D(1, temp)
+            if fmax_lam >= wmin and fmax_lam <= wmax:
+                w = fmax_lam
+                flux_ref = sp(w)
+            elif fmax_lam > wmax:
+                w = wmax
+                flux_ref = yz[np.argmax(xz)]
+            else:
+                w = wmin
+                flux_ref = yz[np.argmin(xz)]
+
+            amp_guess = flux_ref / bb(w)
+            return amp_guess / nbb
+
+        loop_over_non_fixed("dust_continuum", "tau", dust_continuum_guess)
+
+        def line_fwhm_guess(row):
+            w = row["wavelength"][0]
+            if not instrument.within_segment(w, inst):
+                return 0
+
+            fwhm = instrument.fwhm(inst, w, as_bounded=True)[0][0]
+            return fwhm
+
+        def amp_guess(row, fwhm):
+            w = row["wavelength"][0]
+            if not instrument.within_segment(w, inst):
+                return 0
+
+            factor = 1.5
+            wmin = w - factor * fwhm
+            wmax = w + factor * fwhm
+            xz_window = np.logical_and(wmin < xz, xz < wmax)
+            xpoints = xz[xz_window]
+            ypoints = yz[xz_window]
+            if np.count_nonzero(xz_window) >= 2:
+                # difference between flux in window and flux around it
+                power_guess = integrate.trapezoid(yz[xz_window], xz[xz_window])
```

Comment:
Not sure if this may be too heavy.  But people doing lots of fits can of course avoid guessing (or use a simpler, guess(simple_scale = True) thing we develop later).